crates/matrix-sdk-ui/tests/integration/axum.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    net::{Ipv4Addr, SocketAddr},
+    sync::{Arc, Mutex},
+};
+
+use axum::{
+    http::StatusCode,
+    response::{IntoResponse, Response},
+    routing::get,
+    Json,
+};
+use axum_extra::response::ErasedJson;
+use matrix_sdk::{
+    config::RequestConfig,
+    matrix_auth::{Session, SessionTokens},
+    Client, ClientBuilder,
+};
+use matrix_sdk_base::SessionMeta;
+use matrix_sdk_test::{test_json, SyncResponseBuilder};
+use ruma::{api::MatrixVersion, device_id, user_id};
+use tokio::spawn;
+
+pub fn test_client_builder(routes: axum::Router) -> ClientBuilder {
+    let addr = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), 0);
+    let server = axum::Server::bind(&addr).serve(routes.into_make_service());
+
+    let client_builder = Client::builder()
+        .homeserver_url(format!("http://{}", server.local_addr()))
+        .server_versions([MatrixVersion::V1_7]);
+
+    spawn(async move {
+        server.await.unwrap();
```

Comment:
If the server never stops, the spawned task will never stop. Is it what we want?

Reply:
It should be stopped when the tokio runtime is stopped at the end of every test.